### PR TITLE
change multinomial to use async asserts instead of a synchronization

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #134823
* #134819
* __->__ #134818
* #134813

Fixes https://github.com/pytorch/pytorch/issues/134442

cc @malfet

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
gotta fix the tests lol

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -2849,28 +2849,29 @@ def error_inputs_multinomial(op_info, device, **kwargs):
...
-        # error case for the invalid multinomial distribution (sum of probabilities <= 0), 2-D input
-        x = torch.zeros(3, 3, device=device)
-        yield ErrorInput(SampleInput(x, kwargs=kwargs),
-                         error_regex=err_msg2)
-
-        # error case for the invalid multinomial distribution
-        x[1, :] = 1
-        yield ErrorInput(SampleInput(x, kwargs=kwargs),
-                         error_regex=err_msg2)
+    if torch.device(device).type == 'cpu':
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:2852]: These should still error out on cuda. We just need to have a sync within the error catching block right?

**Review by albanD — COMMENTED** [review]
Change sounds good but we need some test right?

**Comment by Chillee** [comment]
@albanD well, the errors are still tested on CPU, which still go through the `aten._assert_async` path.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (Chillee) [reply torch/testing/_internal/common_methods_invocations.py:2852]: hm... it doesn't seem to catch the device-side error message.

```
AssertionError: "invalid multinomial distribution" does not match "CUDA error: device-side assert triggered
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1
Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```

**Pull request closed without merging**